metadata_mapper/validate_mapping.py

```diff
@@ -18,330 +29,143 @@ def solr(**params):
+    except requests.exceptions.HTTPError as err:
         print(
-            f"[{collection_id}]: Validated page {page_path.split('/')[-1]} "
-            f"- {len(collection_report)} errors"
+            f"[Collection {collection_id}]: "
+            f"[{url}]"
+            f"{err}; A valid collection id is required for validation"
         )
+        return
+
+    mapper = collection_data.get("rikolti_mapper_type")
+    mapped_mapper = MAPPER_MAP.get(mapper, mapper)
+    vernacular = utilities.import_vernacular_reader(mapped_mapper)
+
+    return vernacular.record_cls.validator
+
+
+def validate_collection(collection_id: int,
+                        validator_class: Type[Validator] = None,
+                        validator: Validator = None
+                        ) -> Validator:
+    """
+    Validates all pages of a collection of mapped data.
+
+    Parameters:
+        collection_id: int
+            The collection ID
+        validator_class: Type[Validator] (default: None)
+            The validator class to use. Can be derived if not provided.
+        validator: Validator (default: None)
+            A validator instance to use. Can be derived if not provided.
+
+    Returns: Validator
+        The validator containing errors
+    """
+    if not validator_class and not validator:
+        validator_class = get_validator_class(collection_id)
+
+    if not validator:
+        validator = validator_class()
+
+    for page_id in utilities.get_files("mapped_metadata", collection_id):
+        validate_page(collection_id, page_id, validator)
+
+    return validator
+
+
+def validate_page(collection_id: int, page_id: int,
+                  validator: Validator) -> Validator:
+    """
+    Validates a provided page of a provided collection of mapped data.
+
+    Parameters:
+        collection_id: int
+            The collection ID
+        page_id: int
+            The page number within the collection
+        validator: Validator
+            The validator instance to use
+
+    Returns Validator
+        The validator containing validation errors
+    """
+    context = {
+        "collection_id": collection_id,
+        "page_id": page_id
+    }
+    mapped_metadata = validator.generate_keys(
+                        get_mapped_data(collection_id, page_id),
+                        type="Rikolti",
+                        context=context
+                      )
+    solr_data = validator.generate_keys(
+                    get_solr_data(collection_id, list(mapped_metadata.keys())),
+                    type="Solr",
+                    context=context
+                )
+
+    if len(mapped_metadata) == 0 or len(solr_data) == 0:
+        print("No data found in "
+              f"{'mapped Rikolti data' if len(mapped_metadata) else 'Solr'}."
+              " Aborting."
+              )
+        return
 
-    return collection_report
+    intersection = set(mapped_metadata.keys()) & set(solr_data.keys())
```

Comment:
If `len(intersection) != len(mapped_metadata) != len(solr_data)`, then we should probably report out the set difference as well. Given that  `solr_data` is a function of mapped_metadata, it seems quite impossible that `len(solr_data) > len(mapped_metadata)`, so this is likely always a set difference of `set(mapped_metadata.keys()) - set(solr_data.keys())`.

Which also brings up the point that we do need to report out cases where a record exists in Solr that does not exist in Rikolti. Perhaps this is not part of the validator, since it is less a record validation and more a collection extent validation? Where would such a report fit into the framework you've presented here, though?

Reply:
I think such a report could fit, if we wanted to still create a validation CSV and have rows for "Record Doesn't Exist" (I'm thinking a single row per record, not one for each key).

Or, we could optionally generate a separate "errors.csv" or something if this occurs, so it doesn't get mixed up in validation data.

Reply:
Instead of running an intersection, we now `union` the sets to get a list of all keys. If one of the datasets is missing, the validator will produce a single error and bail out.